# STAT 345: Nonparametric Statistics

## Lesson 06.1: The Conover Squared-Ranks Test for Equal Variances

**Reading: Conover Section 5.3**

*Prof. John T. Whelan*

Thursday 27 February 2025

These lecture slides are in a computational notebook.  You have access to them through http://vmware.rit.edu/

Flat HTML and slideshow versions are also in MyCourses.

The notebook can run Python commands (other notebooks can use R or Julia; "Ju-Pyt-R").  Think: computational data analysis, not "coding".

Standard commands to activate inline interface and import libraries:

In [1]:
%matplotlib inline

In [2]:
import numpy as np
from scipy import stats
import matplotlib.pyplot as plt
plt.rcParams['figure.figsize'] = (8.0,5.0)
plt.rcParams['font.size'] = 14

- So far, we've considered rank-based tests of whether two or more samples could have been drawn from distributions w/the same location parameter (median or mean).

- Now consider tests for the scale parameter (variance or the like).

- Given two independent samples, $\{x_i|i=1,\ldots,n\}$ & $\{y_j|j=1,\ldots,m\}$, $H_0$ says the distributions equally spread out.  $H_1$ could be one- or two-sided.

- Recall: parametric $F$-test looks at ratio of sample variances
$$  s_x^2 = \frac{1}{n-1}\sum_{i=1}^n (x_i-{{\overline{x}}})^2 \qquad\hbox{and}\qquad
    s_y^2 = \frac{1}{m-1}\sum_{j=1}^m (y_j-{{\overline{y}}})^2
$$
where ${{\overline{x}}}=\frac{1}{n}\sum_{i=1}^n x_i$ & ${{\overline{y}}}=\frac{1}{m}\sum_{j=1}^m y_j$; statistic is $F = \frac{s_x^2}{s_y^2}$

- For rank-based nonparametric test, look at ranks of $U_i^2 = (x_i-{{\overline{x}}})^2$ and $V_j^2 = (y_j-{{\overline{y}}})^2$ instead of just averaging those quantites.

- Take all of the $\{U_i^2=(x_i-{{\overline{x}}})^2\}$ and $\{V_j^2=(y_j-{{\overline{y}}})^2\}$, rank them in order, and note the ranks $\{R(U_i^2)\}$ and $\{R(V_j^2)\}$.
- An equivalent but easier computation is to take the square roots
$U_i = {\left\lvert x_i-{{\overline{x}}}\right\rvert}$ & $V_j = {\left\lvert y_j-{{\overline{y}}}\right\rvert}$
and rank those instead. Since the mapping from $U_i\rightarrow U_i^2$ is
monotonic, it won’t change the order, and thus we’ll get the exact same
set of ranks if we use $\{R(U_i)\}$ and $\{R(V_j)\}$.

- Consider the example we introduced previously for the Wilcoxon-Mann-Whitney rank sum test:

In [3]:
x_i = np.array([8.56, 5.03, 48.1, 1.31, 4.82]); y_j = np.array([15.0, 12.3, 28.0, 13.9])
n = len(x_i); m = len(y_j); N = n+m; xbar = np.mean(x_i); ybar = np.mean(y_j);
U_i = np.abs(x_i - xbar); V_j = np.abs(y_j - ybar); U_i, V_j

(array([ 5.004,  8.534, 34.536, 12.254,  8.744]),
 array([ 2.3,  5. , 10.7,  3.4]))

In [4]:
UV_r = np.concatenate((U_i,V_j)); R_r = stats.rankdata(UV_r);
R_r, stats.rankdata(np.concatenate(((x_i-xbar)**2,(y_j-ybar)**2)))

(array([4., 5., 9., 8., 6., 1., 3., 7., 2.]),
 array([4., 5., 9., 8., 6., 1., 3., 7., 2.]))

In [5]:
RU_i = R_r[:n]; RV_j = R_r[n:]; RU_i, RV_j

(array([4., 5., 9., 8., 6.]), array([1., 3., 7., 2.]))

- The obvious thing would be to sum the ranks of the $\{U_i\}$, but it turns out you generally get a more powerful test if you square the ranks
before adding them, so the test statistic is
$$T_u = \sum_{i=1}^{n} \bigl(R(U_i)\bigr)^2$$

- Not obvious this is the best thing to do, but sort of analogous to sum of squares in parametric tests.  You'll investigate on the homework and see empirically that $\sum_{i=1}^{n} \bigl(R(U_i)\bigr)^2$ tends
to lead to a more powerful test statistic than $\sum_{i=1}^{n} R(U_i)$.

- Note that
$$ T_u + T_v =
\sum_{i=1}^{n} \bigl(R(U_i)\bigr)^2 + \sum_{j=1}^{m} \bigl(R(V_j)\bigr)^2
=\sum_{r=1}^N R_r^2 = N \overline{R^2}
$$
is determined only by the ranks $\{R_r\}$ and not how they're partitioned between the $U$s and $V$s, so
$T_u$ carries the same information as $T_v$, if you know $\overline{R^2}$.

- If there are no ties, the ranks are just $\{1,\ldots,N\}$ and something called Faulhaber's formula tells is $T_u+T_v=\sum_{r=1}^N R_r^2 = N \overline{R^2}=\frac{N(N+1)(2N+1)}{6}$.

In [6]:
np.sum(R_r**2), N*(N+1)*(2*N+1)//6

(285.0, 285)

- If there are ties, we assign the average ranks in the usual way, but while $\sum_{r=1}^N R_r$ doesn't change if there are ties, $\sum_{r=1}^N R_r^2$ does.  E.g., $2.5+2.5 = 5 = 2+3$ but
$$(2.5)^2 + (2.5)^2 = 6.25 + 6.25 = 12.5 \ne 2^2 + 3^2 = 4 + 9 = 13$$

- Note that $E({\color{royalblue}{T_u}})=n\overline{R^2}$, $E({\color{royalblue}{T_v}})=m\overline{R^2}$, and
$$
T_u - n\overline{R^2} = \frac{(N-n)T_u-nT_v}{N} = \frac{mT_u-nT_v}{N} = -(T_v - m\overline{R^2})
$$
so $T_u - n\overline{R^2}$ will have zero expectation value, and treat the two samples the same way.

- You might reasonably complain that $E({\color{royalblue}{T_u}})$ should be a fixed number, but like the variance of other rank-based statistics, it depends on the observed ranks, specifically on how many ties there are and where. 

- The “dirty secret” of methods like this is that this is actually a conditional expectation value $E({\color{royalblue}{T_u}}|\overline{R^2})$, conditional upon the full set of $N$ available ranks with the actual observed ties taken into account.

- I.e.,
our statements about expectations, significance, power, etc, all pertain
to repeating the experiment but only considering repetitions that have
this particular set of ranks including ties.

- This is kind of
unsatisfying from a frequentist point of view, but it’s necessary to be
able to describe the properties of the test in a distribution-free way,
since you’d have to know the underlying distribution to know how likely
ties are. Note that this is no big deal to a Bayesian, who is always
constructing probabilities for unknown quantities conditional upon the
data which were actually observed.

- Given that $T_u - n\overline{R^2}=-(T_v - m\overline{R^2})$, not surprising that variance treats the samples symmetrically:
$$\operatorname{Var}(T_u - n\overline{R^2})
= \frac{nm}{N-1}\overline{\bigl(R^2-\overline{R^2}\bigr)^2}
= \frac{nm}{N-1}\left(\overline{R^4}-\bigl(\overline{R^2}\bigr)^2\right)$$
where $$\overline{R^4}
  = \frac{1}{N}
  \left(
    \sum_{i=1}^n \bigl(R(U_i)\bigr)^4 + \sum_{j=1}^m \bigl(R(V_j)\bigr)^4
  \right)$$ which is again equal to a constant
($\frac{1}{N}\sum_{r=1}^N r^4=\frac{(N+1)(2N+1)(3N^2+3N-1)}{30}$) if
there are no ties.

The statistic
$$
T_1 = \frac{T_u - n\overline{R^2}}{
    \sqrt{
      \frac{nm}{N-1}\left(\overline{R^4}-\bigl(\overline{R^2}\bigr)^2\right)
    }
  } = \frac{T_u - n\overline{R^2}}{
    \sqrt{
      \frac{nm}{N-1}\overline{\left(R^2-\overline{R^2}\right)^2}
    }
  }
$$
is approximately standard normal (under $H_0$) for large samples,
and can be used to construct hypothesis tests, $p$-values, etc.

In [7]:
Rsq_r = R_r**2; Rsq_r

array([16., 25., 81., 64., 36.,  1.,  9., 49.,  4.])

Since there are no ties in this example, $\overline{R^2}=\frac{(N+1)(2N+1)}{6}$:

In [8]:
Rsqbar = np.mean(Rsq_r); Rsqbar, (N+1)*(2*N+1)/6.

(31.666666666666668, 31.666666666666668)

In [9]:
TU = np.sum(RU_i**2); TV = np.sum(RV_j**2); TU, TV

(222.0, 63.0)

As noted above, $T_u-m\overline{R^2}=-\left(T_v-n\overline{R^2}\right)$:

In [10]:
TU-n*Rsqbar, TV-m*Rsqbar

(63.66666666666666, -63.66666666666667)

In [11]:
varT = n*m/(N-1) * np.mean((Rsq_r-Rsqbar)**2); varT

1752.222222222222

We construct the statistic which should be approximately standard normal, and find it is at about the 93.6th percentile of the null distribution:

In [12]:
T1 = (TU-n*Rsqbar)/np.sqrt(varT); T1

1.5209590473083399

In [13]:
stats.norm.cdf(T1)

0.9358649426940304

This would correspond to a two-sided $p$-value of about $0.13$:

In [14]:
2*stats.norm.sf(T1)

0.12827011461193927

## Exact Null Distribution

- If there are no ties, we can use the exact distribution of the statistic under the null hypothesis.

- Similar to Wilcoxon rank-sum case; $H_0$ says any combination of $n$ out of the $N$ available ranks $1,\ldots,N$ can be the $R(U_i)$.

- Because we square the ranks before adding them, the null distribution is not the same as for the rank sum statistic.  Conover lists some percentiles of the null distribution in Table A9.

- It's not available in Python or R, but we can compute it by making a list of all the possible combinations and constructing the squared ranks statistic for each of them.  (We could have done this for the rank sum statistic too, but we didn't bother because it was available in the software.)

- We use the function `itertools.combinations()` from the `itertools` library to loop over the possible combinations of ranks.  (Note we only need the combinations and not the permutations because each of the $n!$ permutations corresponding to a given combination gives the same value for the statistic.)

In [15]:
import itertools
itertools.combinations(range(1,N+1),n)

The function actually returns an "iterator" which is something you can use inside a loop or similar structure.

In [16]:
for thisRU_i in itertools.combinations(range(1,N+1),n):
    print(thisRU_i)

(1, 2, 3, 4, 5)
(1, 2, 3, 4, 6)
(1, 2, 3, 4, 7)
(1, 2, 3, 4, 8)
(1, 2, 3, 4, 9)
(1, 2, 3, 5, 6)
(1, 2, 3, 5, 7)
(1, 2, 3, 5, 8)
(1, 2, 3, 5, 9)
(1, 2, 3, 6, 7)
(1, 2, 3, 6, 8)
(1, 2, 3, 6, 9)
(1, 2, 3, 7, 8)
(1, 2, 3, 7, 9)
(1, 2, 3, 8, 9)
(1, 2, 4, 5, 6)
(1, 2, 4, 5, 7)
(1, 2, 4, 5, 8)
(1, 2, 4, 5, 9)
(1, 2, 4, 6, 7)
(1, 2, 4, 6, 8)
(1, 2, 4, 6, 9)
(1, 2, 4, 7, 8)
(1, 2, 4, 7, 9)
(1, 2, 4, 8, 9)
(1, 2, 5, 6, 7)
(1, 2, 5, 6, 8)
(1, 2, 5, 6, 9)
(1, 2, 5, 7, 8)
(1, 2, 5, 7, 9)
(1, 2, 5, 8, 9)
(1, 2, 6, 7, 8)
(1, 2, 6, 7, 9)
(1, 2, 6, 8, 9)
(1, 2, 7, 8, 9)
(1, 3, 4, 5, 6)
(1, 3, 4, 5, 7)
(1, 3, 4, 5, 8)
(1, 3, 4, 5, 9)
(1, 3, 4, 6, 7)
(1, 3, 4, 6, 8)
(1, 3, 4, 6, 9)
(1, 3, 4, 7, 8)
(1, 3, 4, 7, 9)
(1, 3, 4, 8, 9)
(1, 3, 5, 6, 7)
(1, 3, 5, 6, 8)
(1, 3, 5, 6, 9)
(1, 3, 5, 7, 8)
(1, 3, 5, 7, 9)
(1, 3, 5, 8, 9)
(1, 3, 6, 7, 8)
(1, 3, 6, 7, 9)
(1, 3, 6, 8, 9)
(1, 3, 7, 8, 9)
(1, 4, 5, 6, 7)
(1, 4, 5, 6, 8)
(1, 4, 5, 6, 9)
(1, 4, 5, 7, 8)
(1, 4, 5, 7, 9)
(1, 4, 5, 8, 9)
(1, 4, 6, 7, 8)
(1, 4, 6

You make an array of the values it loops over by using a list comprehension:

In [17]:
RU_Ii = np.array([thisRU_i for thisRU_i in itertools.combinations(range(1,N+1),n)]); RU_Ii.shape

(126, 5)

There are $\binom{N}{n}=\frac{N!}{n!m!}$ equally likely combinations, which in this case is $126$:

In [18]:
from scipy.special import comb as nchoosek
nchoosek(N,n)

126.0

So for each of these combinations we construct the statistic $T_u$ and get a list of equally likely values:

In [19]:
TU_I = np.sort(np.sum(RU_Ii**2,axis=-1)); TU_I

array([ 55,  66,  75,  79,  82,  87,  88,  90,  94,  95,  99, 100, 103,
       103, 106, 110, 111, 111, 114, 114, 115, 115, 118, 120, 120, 121,
       123, 126, 127, 127, 127, 129, 130, 130, 131, 132, 134, 135, 135,
       135, 138, 138, 139, 142, 142, 143, 143, 144, 145, 146, 147, 148,
       150, 151, 151, 152, 154, 155, 155, 156, 158, 159, 159, 159, 159,
       160, 162, 162, 163, 165, 166, 166, 167, 168, 169, 171, 171, 172,
       174, 174, 175, 175, 175, 176, 178, 179, 180, 180, 183, 183, 183,
       186, 186, 187, 190, 190, 191, 191, 192, 194, 195, 195, 198, 199,
       200, 201, 204, 206, 207, 207, 207, 210, 211, 214, 215, 219, 220,
       222, 223, 228, 231, 234, 235, 239, 246, 255])

Recall the actual value of $T_u$ for reference:

In [20]:
TU

222.0

We can calculate the cdf and $p$-value and see that the normal approximation ($p=0.13$) was not so far off:

In [21]:
np.mean(TU_I <= TU)

0.9365079365079365

In [22]:
2*np.mean(TU_I >= TU)

0.14285714285714285

We can also see from the list that there are 9 values of the statistic equal to or greater than 222.

In [23]:
np.sum(TU_I >= TU)

9

In [24]:
TU_I[TU_I >= TU]

array([222, 223, 228, 231, 234, 235, 239, 246, 255])

## Extension to $k$-Sample Case

- The squared ranks test can be extended to the case of $k$ independent samples, just as the Kruskal-Wallis test is defined as the $k$-sample generalization of the Wilcoxon rank-sum test.

- The test statistic is written in terms of the squared ranks
$S_i = \sum_{j=1}^{n_i} \bigl(R({\lvert x_{ij}-{{\overline{x}}}_i\rvert})\bigr)^2$
as $$\frac{\sum_{i=1}^k \frac{S_i^2}{n_i} - N\left(\overline{R^2}\right)^2}
  {
    \frac{N}{N-1}
    \left(
      \overline{R^4}-N\bigl(\overline{R^2}\bigr)^2
    \right)
  }=\frac{\sum_{i=1}^k \frac{1}{n_i}\bigl(S_i - n_i\overline{R^2}\bigr)^2}
  {
    \frac{N}{N-1}
    \overline{\bigl(R^2-\overline{R^2}\bigr)^2}
  }$$
where $\overline{R^2}=\frac{1}{N}\sum_{r=1}^N R_r^2$ is also called $\overline{S}$

- The null distribution is approximately $\chi^2(k-1)$.

Look at the example from last time:

In [25]:
x_i_j = [np.array([ 14.97,   5.80,  25.03,   5.50 ]),
       np.array([  5.83,  13.96,  21.96]),
       np.array([ 17.89,  23.03,  61.09,   18.62,  55.51])]
n_i = np.array([len(xi_j) for xi_j in x_i_j]); k = len(n_i); N = np.sum(n_i); n_i,k,N

(array([4, 3, 5]), 3, 12)

In [26]:
xbar_i = np.array([np.mean(xi_j) for xi_j in x_i_j]); xbar_i

array([12.825     , 13.91666667, 35.228     ])

In [27]:
U_i_j = [np.abs(xi_j-np.mean(xi_j)) for xi_j in x_i_j]; U_i_j

[array([ 2.145,  7.025, 12.205,  7.325]),
 array([8.08666667, 0.04333333, 8.04333333]),
 array([17.338, 12.198, 25.862, 16.608, 20.282])]

In [28]:
U_r = np.concatenate(U_i_j); RU_r = stats.rankdata(U_r)
i_r = np.concatenate([(i,)*n_i[i] for i in range(k)])
RU_i_j = [RU_r[i_r==i] for i in range(k)]; RU_i_j

[array([2., 3., 8., 4.]),
 array([6., 1., 5.]),
 array([10.,  7., 12.,  9., 11.])]

Now construct the sums of the squared ranks $S_i = \sum_{j=1}^{n_i} \bigl(R({\lvert x_{ij}-{{\overline{x}}}_i\rvert})\bigr)^2$

In [29]:
S_i = np.array([np.sum(RUi_j**2) for RUi_j in RU_i_j]); S_i

array([ 93.,  62., 495.])

The average squared rank $\overline{S}=\frac{1}{N}\sum_{r=1}^N R_r^2$:

In [30]:
Sbar = np.mean(RU_r**2); Sbar

54.166666666666664

The numerator $\sum_{i=1}^k \frac{S_i^2}{n_i} - N\left(\overline{R^2}\right)^2=\sum_{i=1}^k \frac{1}{n_i}\bigl(S_i - n_i\overline{R^2}\bigr)^2$

In [31]:
np.sum(S_i**2/n_i)-N*Sbar**2, np.sum((S_i-n_i*Sbar)**2/n_i)

(17240.250000000007, 17240.25)

The denominator $D^2=
    \frac{N}{N-1}
    \overline{\bigl(R^2-\overline{R^2}\bigr)^2}=    \frac{N}{N-1}
\left(
      \overline{R^4}-\bigl(\overline{R^2}\bigr)^2
    \right)$

In [32]:
Dsq=N/(N-1)*np.mean((RU_r**2-Sbar)**2); Dsq, N/(N-1)*(np.mean(RU_r**4)-Sbar**2),(np.sum(RU_r**4)-N*Sbar**2)/(N-1) 

(2318.3333333333335, 2318.333333333334, 2318.333333333334)

Tip: it's easy to mess up the cancellation in the "shortcut formula" versions:

In [33]:
np.sum(RU_r**4-N*Sbar**2)/(N-1)

-32889.99999999999

Obviously wrong since $D^2>0$ by construction.  Can avoid this by:

- Sanity-checking the values of intermediate quantities

- Using versions which square differences, rather than subtracting squares (also avoids roundoff)

Now assemble the statistic

In [34]:
T = np.sum((S_i-n_i*Sbar)**2/n_i)/Dsq; T

7.436484543493889

Finally, get the $p$-value from the upper tail (only!) of the $\chi^2(k-1)$ distribution:

In [35]:
stats.chi2(df=k-1).sf(T)

0.024276602034339175

## More tips on the multinomial iterator from problem set 05.1

#### Problem 5.2.2:

Find the exact distribution of the Kruskal-Wallis test statistic when
$H_0$ is true, $n_1=3$, $n_2=2$, $n_3=1$, and there are no ties. Compare
your results with the quantiles given in Table A8.
\[Also compare with the chi-squared approximation.\]

I pointed to a function called `multinomial_combinations`, which is now part of the `combinatorics` library available from https://phillipmfeldman.org/Python/combinatorics.html

In [36]:
from combinatorics import m_way_ordered_combinations as multinomial_combinations

This is an "iterator", which lets you loop through the $\frac{6!}{3!2!1!}=60$ different ways of partitioning the $N=6$ ranks $\{1,2,3,4,5,6\}$ into groups of $n_1=3$, $n_2=2$, and $n_3=1$.

In [37]:
n_i = np.array([3,2,1]); N = np.sum(n_i); k = len(n_i); R_r = np.arange(1,N+1); R_r

array([1, 2, 3, 4, 5, 6])

Each time I access the iterator, it gives me a new example combination of the ranks $R_r$ into $\{R_{1j}\}$, $\{R_{2j}\}$, and $\{R_{3j}\}$, so I could use it to construct $R_1=\sum_{j=1}^{n_1}R_{1j}$, $R_2=\sum_{j=1}^{n_2}R_{2j}$, and $R_3=\sum_{j=1}^{n_3}R_{3j}$ corresponding to that combination.

In [38]:
# Note the new library complains if we pass it n_i as a NumPy array, so we convert it to a list
for R_i_j in multinomial_combinations(R_r,list(n_i)):
    print(R_i_j)
    R1 = np.sum(R_i_j[0])
    R2 = np.sum(R_i_j[1])
    R3 = np.sum(R_i_j[2])
    print(R1,R2,R3)

((1, 2, 3), (4, 5), (6,))
6 9 6
((1, 2, 3), (4, 6), (5,))
6 10 5
((1, 2, 3), (5, 6), (4,))
6 11 4
((1, 2, 4), (3, 5), (6,))
7 8 6
((1, 2, 4), (3, 6), (5,))
7 9 5
((1, 2, 4), (5, 6), (3,))
7 11 3
((1, 2, 5), (3, 4), (6,))
8 7 6
((1, 2, 5), (3, 6), (4,))
8 9 4
((1, 2, 5), (4, 6), (3,))
8 10 3
((1, 2, 6), (3, 4), (5,))
9 7 5
((1, 2, 6), (3, 5), (4,))
9 8 4
((1, 2, 6), (4, 5), (3,))
9 9 3
((1, 3, 4), (2, 5), (6,))
8 7 6
((1, 3, 4), (2, 6), (5,))
8 8 5
((1, 3, 4), (5, 6), (2,))
8 11 2
((1, 3, 5), (2, 4), (6,))
9 6 6
((1, 3, 5), (2, 6), (4,))
9 8 4
((1, 3, 5), (4, 6), (2,))
9 10 2
((1, 3, 6), (2, 4), (5,))
10 6 5
((1, 3, 6), (2, 5), (4,))
10 7 4
((1, 3, 6), (4, 5), (2,))
10 9 2
((1, 4, 5), (2, 3), (6,))
10 5 6
((1, 4, 5), (2, 6), (3,))
10 8 3
((1, 4, 5), (3, 6), (2,))
10 9 2
((1, 4, 6), (2, 3), (5,))
11 5 5
((1, 4, 6), (2, 5), (3,))
11 7 3
((1, 4, 6), (3, 5), (2,))
11 8 2
((1, 5, 6), (2, 3), (4,))
12 5 4
((1, 5, 6), (2, 4), (3,))
12 6 3
((1, 5, 6), (3, 4), (2,))
12 7 2
((2, 3, 4), (1, 5), (6

You can either do a loop like the one above, and compute another equally likely value of the Kruskal-Wallis statistic for each one, or you can combine them all in a list like this:

In [39]:
R_I_i_j = [R_i_j for R_i_j in multinomial_combinations(R_r,list(n_i))]

There are clever ways to get the array of rank-sums $R^{(I)}_i$ from this, but a little more brute-force is just to pull out the $R^{(I)}_{1j}$:

In [40]:
R1_Ij = np.array([R_i_j[0] for R_i_j in R_I_i_j]); R1_Ij.shape

(60, 3)

and then compute $R^{(I)}_1=\sum_{j=1}^{n_1}R^{(I)}_{1j}$:

In [41]:
R1_I = R1_Ij.sum(axis=-1); R1_I

array([ 6,  6,  6,  7,  7,  7,  8,  8,  8,  9,  9,  9,  8,  8,  8,  9,  9,
        9, 10, 10, 10, 10, 10, 10, 11, 11, 11, 12, 12, 12,  9,  9,  9, 10,
       10, 10, 11, 11, 11, 11, 11, 11, 12, 12, 12, 13, 13, 13, 12, 12, 12,
       13, 13, 13, 14, 14, 14, 15, 15, 15])

We can do likewise for $R^{(I)}_2=\sum_{j=1}^{n_2}R^{(I)}_{2j}$ and $R^{(I)}_3=\sum_{j=1}^{n_3}R^{(I)}_{3j}$:

In [42]:
R2_Ij = np.array([R_i_j[1] for R_i_j in R_I_i_j]); R2_Ij.shape

(60, 2)

In [43]:
R2_I = R2_Ij.sum(axis=-1); R2_I

array([ 9, 10, 11,  8,  9, 11,  7,  9, 10,  7,  8,  9,  7,  8, 11,  6,  8,
       10,  6,  7,  9,  5,  8,  9,  5,  7,  8,  5,  6,  7,  6,  7, 11,  5,
        7, 10,  5,  6,  9,  4,  7,  9,  4,  6,  8,  4,  5,  7,  3,  7,  8,
        3,  6,  7,  3,  5,  6,  3,  4,  5])

In [44]:
R3_Ij = np.array([R_i_j[2] for R_i_j in R_I_i_j]); R3_Ij.shape

(60, 1)

In [45]:
R3_I = R3_Ij.sum(axis=-1); R3_I

array([6, 5, 4, 6, 5, 3, 6, 4, 3, 5, 4, 3, 6, 5, 2, 6, 4, 2, 5, 4, 2, 6,
       3, 2, 5, 3, 2, 4, 3, 2, 6, 5, 1, 6, 4, 1, 5, 4, 1, 6, 3, 1, 5, 3,
       1, 4, 3, 1, 6, 2, 1, 5, 2, 1, 4, 2, 1, 3, 2, 1])

We can manipulate these, and check, for example, that $R_1+R_2+R_2=\frac{N(N+1)}{2}=21$ for each of the possible combinations:

In [46]:
R1_I + R2_I + R3_I

array([21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21,
       21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21,
       21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21,
       21, 21, 21, 21, 21, 21, 21, 21, 21])

Given these three arrays of $R^{(I)}_1$, $R^{(I)}_2$, and $R^{(I)}_3$, we can construct the 60 equally-likely Kruskal-Wallis statistics $T^{(I)}$.